Simplify, clean up, fix a few things

mrd0ll4r opened the pull request:
Alright, here's the PR. What this does:

- Use `bool` instead of `u8` for most of the single-bit binary values. I don't use the interrupt stuff on my device, so I didn't mess with that much.
- Fix `PinMode` constants being mixed up (fixes #2 )
- Fix `write_bit` to work correctly with the `HIGH` and `LOW` constants (fixes #3 )
- Improve documentation.
- Remove dependencies on everything but `embedded_hal`. This also removes the `unsafe` code needed for generic arrays.
- Rework `read_` and `write_` functions to have one variant for single registers and one for double registers each.
- Remove unused functions (`init_hardware` and maybe others too)
- Cargo fmt :)

It's a big one, and it's breaking, so... use at your own discretion :)

EDIT: I can squash these into one commit if you prefer

sentinelt commented on src/lib.rs:
```diff
@@ -156,24 +131,29 @@ where
-        Ok((u16(buffer[0]) << 8) + u16(buffer[1]))
+    pub fn read_gpioab(&mut self) -> Result<u16, E> {
+        let buffer = self.read_double_register(Register::GPIOA)?;
+        Ok((buffer[0] as u16) << 8 + (buffer[1] as u16))
```
You removed necessary parentheses, this line should be:
Ok(((buffer[0] as u16) << 8) + (buffer[1] as u16))

but due to the Rust operators precedence rules your version is interpreted as:
Ok((buffer[0] as u16) << (8 + (buffer[1] as u16)))

Even better I propose to change this line to:
Ok((buffer[0] as u16) << 8 | (buffer[1] as u16))

mrd0ll4r replied:
Oh wow, I was unaware, thank you!

eminence commented:
Hi all, just chiming in to let you all know that I've been running with a3d072754abca60a92ece820f7cfb767a0c11669 for several weeks, and it's been solid for me

lucazulian commented:
> Hi all, just chiming in to let you all know that I've been running with [a3d0727](https://github.com/lucazulian/mcp23017/commit/a3d072754abca60a92ece820f7cfb767a0c11669) for several weeks, and it's been solid for me

Can you give a try to the latest version so I can merge this PR? Thank you!

eminence commented:
Yup, I'm now running on the latest version, but my code doesn't use the `read_gpioab` function (so it won't be able to test the latest fix to that)

mrd0ll4r commented:
bump :)

jordens commented:
@mrd0ll4r @lucazulian ping
We've been using this for a while in [production](https://github.com/quartiq/stabilizer) now with great success.
It would be great to see this merged and released.

lucazulian commented:
> @mrd0ll4r @lucazulian ping
> We've been using this for a while in [production](https://github.com/quartiq/stabilizer) now with great success.
> It would be great to see this merged and released.

Wow, that's incredible! I'm so sorry for the delay, I check immediately!

jordens commented:
Thanks, also for the very useful crate!

lucazulian commented:
> Thanks, also for the very useful crate!

Can you resolve conflicts (use your branch)? Thank you

mrd0ll4r commented:
I'll see what I can do

mrd0ll4r commented:
I rebased this on top of the current master branch. For some reason this PR used to include PR #1, which was merged. I apparently rebased on top of that back then, but it doesn't show up in the commit history for your master anymore. So after rebasing on top of master again now it's gone from this PR.
I removed the `init_hardware` function with this PR as it was never called. I'm not sure what happened over the last year, so maybe make sure that's still correct. FWIW, I've (and apparently others) been running fine without it for quite some time now :)
Also, I'd recommend cutting a new release, since these are breaking changes.
Cheers!

lucazulian commented:
@mrd0ll4r than you for you effort! I will re-introduce `init_hardware` as it's present in the repos I used, but it's a user choice to use or not.
